geps/gep-2162.md

```diff
@@ -0,0 +1,110 @@
+
+Partnered with the [Conformance Profiles](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-1709.md) work, we want to:
+
+1. Enable users to easily see what features an implementation supports
+1. Improve UX and potentially provide clear errors when a user is trying to use an unsupported feature
+
+This doc proposes to enhance the GatewayClassStatus API so implementations could publish a list of features they support/don't support.
+
+Implementations **should** publish the supported/unsupported features before Accepting the GatewayClass.
+
+They can either maintain static lists of supported/unsupported features OR generate them from their conformance tests results/reports as part of their CI.
+
+## API
+
+This GEP proposes API changes describes as follow:
+1. Adding `FeaturesSupport` and  `GatewayClassFeatures` structs. 
+2. Update the `GatewayClassStatus` struct to include `GatewayClassFeatures`
+
+```go
+// FeaturesSupport contains the supported features and unsupported features by the installed GatewayClass.
+type FeaturesSupport struct {
+	SupportedFeatures []string `json:"supportedFeatures,omitempty"`
+
+	UnsupportedFeatures []string `json:"unsupportedFeatures,omitempty"`
+}
+
+// GatewayClassFeatures is the features supported by the GatewayClass broken down by support level.
+type GatewayClassFeatures struct {
+	// Core indicates the core support level features. UnsupportedFeatures should be null for the implementation to be considered all conformant. 
+    Core FeaturesSupport `json:"core,omitempty"`
+
+	// Extended indicates the extended support level features.
+    Extended FeaturesSupport `json:"extended,omitempty"`
+    
+}
+// GatewayClassStatus is the current status for the GatewayClass.
+type GatewayClassStatus struct {
+	// Conditions is the current status from the controller for
+	// this GatewayClass.
+	//
+	// Controllers should prefer to publish conditions using values
+	// of GatewayClassConditionType for the type of each Condition.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{type: "Accepted", status: "Unknown", message: "Waiting for controller", reason: "Pending", lastTransitionTime: "1970-01-01T00:00:00Z"}}
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+    // Features is the features supported by the GatewayClass broken down by support level.
+    Features GatewayClassFeatures `json:"features,omitempty"`
+}
+```
+
+## Conformance Details
+
+(TBD)
+
+## Alternatives
+
+### Re-using ConformanceProfiles structs
+
+We could use the same structs as we do in conformance profile object, more specifically, the [ProfileReport](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/apis/v1alpha1/profilereport.go#LL24C6-L24C19) struct.
```

Comment:
conformance profiles rely on supported/unsupported features (they are at a higher level of abstraction), hence I think we should use the features for the GWC, as they can be considered as a source of truth.

Reply:
I agree here, it is also related to [Rob's comment](https://github.com/kubernetes-sigs/gateway-api/pull/2163#discussion_r1253441917). IIRC @shaneutt has different ideas in mind but if you both are good with this then I will add it.

Reply:
No I'm basically aligned with @mlavacca here, if we can just have one list of the actual features that can populate here, let's have one unified list.

Reply:
wouldn't it be easier for the conformance profiles to have a list broken down by support level? One unified list would require another small bit of logic to determine support levels.

I am good with either

Reply:
I think I lean towards one unified list. Although I don't expect it to be common, it's possible that some features currently labeled as "Extended" may eventually transition to "Core" or vice versa.

Reply:
> I think I lean towards one unified list. Although I don't expect it to be common, it's possible that some features currently labeled as "Extended" may eventually transition to "Core" or vice versa.

Yeah, it makes sense. I didn't consider it. +1

Reply:
done, PTAL

Reply:
I think not using the conformance structs themselves are fine, but we should put a requirement somewhere in this GEP that the feature names used for the supported features are the same or at least derivative of the conformance test names.

Reply:
I think I'm agreeing with what @youngnick is saying above, but I'll state what I'm thinking just to make sure:

1. `GatewayClass.Status.SupportedFeatures` should use the same names verbatim as `SupportedFeatures` in conformance tests.
2. `SupportedFeatures` in conformance tests should be easy to understand, potentially even mapping to conformance test names.
3. As a result of 1 and 2, we should probably do an audit to be 100% sure we're happy with existing names for supported features because they're about to be part of our API Surface (and thus it will be near impossible to change them going forward).

Does that all line up with what others are thinking here?

Reply:
Agree with all three points here. As part of implementing this, we need to audit all the conformance test names, and probably even standardize a format for feature and/or conformance test names.

Reply:
I have opened https://github.com/kubernetes-sigs/gateway-api/discussions/2285 with some findings - PTAL !